lib/features/steps/runner_steps.rb

```diff
@@ -62,3 +62,65 @@
   $logger.warn 'Sleep was used! Please avoid using sleep in tests!'
   sleep(seconds)
 end
+
+# Starts an interactive terminal
+When('I start a new terminal') do
+  Runner.stop_interactive_session
+  Runner.get_interactive_session
+end
+
+# Stops currently running interactive terminal
+When('I stop the current terminal') do
+  Runner.stop_interactive_session
+end
+
+# Run a command on the terminal
+#
+# @step_input command [String] The command to run on the terminal
+When('I input {string} interactively') do |command|
+  current_terminal = Runner.get_interactive_session
+  success = current_terminal.run_command(command)
+  assert(success, 'The terminal had already closed')
+end
+
+# Get the current buffer values in the terminal
+#
+# @step_input expected_chars [String] The chars present in current buffer
+Then('the terminal is outputting {string}') do |expected_chars|
+  current_terminal = Runner.get_interactive_session
+  assert_equal(expected_chars, current_terminal.current_buffer)
+end
+
+# Verify a string appears in the stdout logs
+#
+# @step_input expected_line [String] The string present in stdout logs
+Then('the terminal has output {string}') do |expected_line|
+  current_terminal = Runner.get_interactive_session
+  match = current_terminal.parsed_output.any? { |line| line == expected_line }
+  assert(match, "No output lines matched #{expected_line}")
+end
+
+# Verify a string appears in the stderr logs
+#
+# @step_input expected_err [String] The string present in stderr logs
+Then('the terminal has the error message {string}') do |expected_err|
+  current_terminal = Runner.get_interactive_session
+  match = current_terminal.parsed_errors.any? { |line| line == expected_err }
+  assert(match, "No output lines matched #{expected_err}")
+end
+
+# Verify the exit code of the terminal
+#
+# @step_input exit_code [Integer] The expected exit code
+Then('the terminal exit code equals {int}') do |exit_code|
+  current_terminal = Runner.get_interactive_session
+  assert_equal(exit_code, current_terminal.last_exit_code)
+end
+
+# Assert that the terminal exited with an error code
+#
+# @step_input exit_code [Integer] The expected exit code
+Then('the terminal exited with an error code') do
```

Comment:
Should we have either of those?  An exit code of 0 for success is only a convention (albeit very common) and there may be other codes that are considered "success".  I think people could make the wrong assumption in certain contexts and it would be safer just to go on explicit exit codes.  Have a step `the exit code is not {int}` to deal with anything non-zero, if that is wanted (which I can see it would be in most practical cases).

Reply:
I think a specific exit code should always have specific text detailing the error too, which is much more understandable to assert against IMO

The only time a specific exit code might be important to check is if we're testing a tool that we wrote, but even then the output is what humans read so seems more useful to check that

I don't think any tool uses 0 for failure or >0 for success as those have been the understood meanings of exit codes for decades (and they'd fail every CI build 😀)

Reply:
I think making an assumption that 0 is a success is reasonable, so long as that's documented and it's possible to look for other codes.

Reply:
👍